crates/util/src/task.rs

```diff
@@ -0,0 +1,129 @@
+use std::future::Future;
+use std::sync::LazyLock;
+
+use tokio_util::sync::CancellationToken;
+use tokio_util::task::TaskTracker;
+
+pub trait FutureOutputExt {
+    fn cancelled() -> Self;
+}
+
+impl FutureOutputExt for () {
+    fn cancelled() -> Self {}
+}
+
+impl<T> FutureOutputExt for anyhow::Result<T> {
+    fn cancelled() -> Self {
+        Err(anyhow::anyhow!("Cancelled due to graceful shutdown"))
+    }
+}
+
+/// Spawns a future on the `tokio` runtime through a
+/// [`tokio_util::task::TaskTracker`]. This ensures that upon graceful shutdown
+/// the future will have already completed or will be cancelled in an orderly
+/// fashion.
+pub fn spawn<F>(future: F) -> tokio::task::JoinHandle<F::Output>
+where
+    F: Future + Send + 'static,
+    F::Output: FutureOutputExt + Send + 'static,
+{
+    let Handle {
+        task_tracker,
+        cancellation_token,
+    } = HANDLE.clone();
+
+    task_tracker.spawn(async move {
+        tokio::select! {
+            _ = cancellation_token.cancelled() => {
+                F::Output::cancelled()
```

Comment:
Here.

Reply:
Otherwise the signature would have to be
```
pub fn spawn<F>(future: F) -> tokio::task::JoinHandle<Option<F::Output>>
```
to return `None` if _gracefully_ cancelled.